### PR TITLE
server: Add telemetry to track hits to /health, /_status/vars

### LINKED ISSUE #40731: ui: add telemetry to track loads of prometheus and health check endpoints
We should start capturing telemetry via segment on the backend when users make requests to the `/_status/vars` and `/health` endpoints. 

Understanding the usage of `/_status/vars` will help us figure out whether metrics are consumed via the prometheus endpoint in place of or in addition to users using the admin UI. 

Understanding usage of `/health` can help us understand whether clusters in the wild are employing load balancers.

Pete V pointed out that these pages have no JS, so it is slightly unclear why we see some data in our telemetry from these endpoints - they should not be able to trigger page loaded events. This telemetry will need to be plumbed via the backend to properly trigger events when the endpoint is accessed.

### PR DESCRIPTION
**Pull request by itsbilal**
Increment telemetry feature counters when the `/health` and
`/_status/vars` endpoints are loaded. The counter emitted is
"monitoring.prometheus.vars" for /_status/vars and
"monitoring.health.details" for /health .

Fixes: #40731

Release note: None

Release justification: Minor monitoring change with the benefit of
tracking feature adoption.

### CODE REVIEW

**Review by rohany — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40926)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @rohany)

___
*[pkg/server/status.go, line 611 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40926#-Lp9nine1hNJ9VnFpjms:-Lp9nine1hNJ9VnFpjmt:b-sw8u9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b3c7823354e06c6bac56cb7cfe773010c67983e/pkg/server/status.go#L611)):*
> ```Go
> 		return nil, grpcstatus.Errorf(codes.InvalidArgument, err.Error())
> 	}
> 	telemetry.Count("web-ui.health")
> ```

I think it’s better to use getCounterOnce to avoid looking up into/locking the counters map repeatedly

**Review by petermattis — COMMENTED** [review]
Looks good on the surface, but I'm not familiar enough with the telemetry patterns to give this a real review.

@dt, @knz, or @solongordon Can one of you give this a look?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40926#-:-LpEAiGU54AXjS16OCUO:bfac3jj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @petermattis)

File: pkg/server/status_test.go [hunk 1]
```diff
@@ -137,6 +137,60 @@ func TestStatusJson(t *testing.T) {
 	}
 }
 
+// TestHealthTelemetry confirms that hits on some status endpoints increment
+// feature telemetry counters.
+func TestHealthTelemetry(t *testing.T) {
+
```
Comment (knz) [inline pkg/server/status_test.go]: nit: remove empty line

File: pkg/server/status_test.go [hunk 2]
```diff
@@ -137,6 +137,60 @@ func TestStatusJson(t *testing.T) {
...
+	defer leaktest.AfterTest(t)()
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(context.TODO())
+
+	rows, _ := db.Query("SELECT * FROM crdb_internal.feature_usage WHERE feature_name LIKE 'web-ui%' AND usage_count > 0;")
```
Comment (knz) [inline pkg/server/status_test.go]: why not check the `err` return here?

File: pkg/server/status_test.go [hunk 3]
```diff
@@ -137,6 +137,60 @@ func TestStatusJson(t *testing.T) {
...
+	if _, err := getText(s, s.AdminURL()+statusPrefix+"vars"); err != nil {
+		t.Fatal(err)
+	}
+
+	expectedCounts := map[string]int{
+		"web-ui.health": 1,
+		"web-ui.prometheus.vars": 1,
+	}
+
+	rows, _ = db.Query("SELECT feature_name, usage_count FROM crdb_internal.feature_usage WHERE feature_name LIKE 'web-ui%' AND usage_count > 0;")
```
Comment (knz) [inline pkg/server/status_test.go]: ditto

File: pkg/server/status_test.go [hunk 4]
```diff
@@ -137,6 +137,60 @@ func TestStatusJson(t *testing.T) {
...
+		}
+
+		if count, ok := expectedCounts[featureName]; ok {
+			if count != usageCount {
+				t.Fatalf("expected %d count for feature %s, got %d", count, featureName, usageCount)
+			}
+			delete(expectedCounts, featureName)
+		}
+	}
+	if err := rows.Close(); err != nil {
```
Comment (knz) [inline pkg/server/status_test.go]: hoist this above just after `Query`'s err test with a `defer`

File: pkg/server/status.go [hunk 5]
```diff
@@ -608,6 +608,7 @@ func (s *statusServer) Details(
 	if err != nil {
 		return nil, grpcstatus.Errorf(codes.InvalidArgument, err.Error())
 	}
+	telemetry.Count("web-ui.health")
```
Comment (knz) [inline pkg/server/status.go]: The name of this feature is misleading. This service method is used for a bunch of things, not just the web UI. As-is, you're going to get over-reporting.

If you want to specialize to web UI access you'll need to detect that the client is the web UI, presumably via a field override in DetailsRequest.

If you're intending to count all accesses to the RPC endpoint, you'll need a better feature name.

File: pkg/server/status.go [hunk 6]
```diff
@@ -1164,6 +1165,7 @@ func (s *statusServer) handleVars(w http.ResponseWriter, r *http.Request) {
 		log.Error(r.Context(), err)
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 	}
+	telemetry.Count("web-ui.prometheus.vars")
```
Comment (knz) [inline pkg/server/status.go]: similar comment here. Many monitoring tools besides crdb's own web UI use this endpoint.

My general recommendation would be to create feature names that describe exactly what's being measured, now how you think this control path is being taken from.

File: pkg/server/status.go (hunk 5 above)
Comment (itsbilal) [reply pkg/server/status.go]: It's intended to be about monitoring in general, so I've updated it to be `monitoring.*` instead of `web-ui.*`. Good point!

**Comment by itsbilal** [comment]
Fixed test failures and addressed comments. PTAL!

**Review by knz — APPROVED** [review]
I recommend you copy paste the name of the two counters in the commit message, so that the next person looking at the commit log needs less time to search for them.

**Comment by itsbilal** [comment]
Done. Thanks for the review @knz!

bors r+

**Pull request merged**